proftpd/utils/ftpshut.8
.TH ftpshut 8 "March 2003"
.\" Process with
.\" groff -man -Tascii ftpshut.8 
.\"
.SH NAME
ftpshut \- shut down all proftpd servers at a given time
.SH SYNOPSIS
.B ftpshut
[
.BI \-R
] [
.BI \-l " min "
] [
.BI \-d " min " 
]
.I time
[
.I "warning-message "
]
.SH DESCRIPTION
The
.BI ftpshut
command provides an automated shutdown procedure that a superuser
can use to automagically disconnect current proftpd connections,
and deny new connections.  The command creates a control file as
.B /etc/shutmsg
to inform proftpd processes about an impending shutdown.
.SH OPTIONS
.TP 10
.B time
The time at which the ftp servers will be brought down.  It may be the
word `now', indicating an immediate shutdown, or a future time in one
of two formats:
.I +number 
or
.I HHMM.  
The first form brings the servers down in
.I number
minutes, while the second form indicates an exact time of day, using
a 24-hour clock format.
.TP
.B -R
Instructs ftpshut to
.I remove
the /etc/shutmsg file, restoring proftpd to its normal service.  If this option
is used, this will be the only operation performed by
.I ftpshut
.
.TP
.B -l min
Specifies the number of minutes before shutdown that new ftp access will
be disabled.  If the -l option is not specified, a default of 10 minutes
(or immediately, if less than 10 minutes remains until shutdown) is used.
.TP
.B -d min
Specifies the number of minutes before shutdown that existing ftp
connections will be disconnected.  If the -d option is not specified,
a default of 5 minutes (or immediately, if less than 5 minutes remains
until shutdown) is used.
.TP
.B warning-message
Optionally provide a message which will be sent to new connections
(if denied) and existing connections (when disconnected).  The message
can contain the following magic cookies, which will be replaced as
follows before being sent to new or existing connections:
.RS
.PD 1
.TP 4
.B %s
time proftpd is being shut down
.PD 0
.TP
.B %r
time new connections will be denied
.TP
.B %d
time current connections will be dropped
.TP
.B %C
current working directory (where applicable)
.TP
.B %L
local host name (or virtualhost name)
.TP
.B %R
remote host name
.TP
.B %T
local time (format Thu Nov 15 17:12:42 1990)
.TP
.B %U
username given at login time
.PD 1
.SH CAVEATS
Proftpd servers can only be shutdown via ftpshut if a master daemon
is running in
.I standalone 
mode.  Servers running in
.I inetd 
mode will honor the shutdown message file created by ftpshut, however
currently running servers may be chroot(2)ed, and unable to read
the shutdown message control file.
.PP
It is possible to set an absolute shutdown time which is 
.B earlier 
than the current time.  If this is the case, proftpd will not shutdown
until the indicated time on the following day.
.SH FILES
.PD 0
.B NONE/sbin/ftpshut
.br
.B /etc/shutmsg
.PD
.SH AUTHORS
.PP
ProFTPD is written and maintained by a number of people, full credits
can be found on
.BR http://www.proftpd.org/credits.html
.PD
.SH SEE ALSO
.BR inetd(8), ftp(1), proftpd(8), ftpwho(1), ftpcount(1)
.PP
Full documentation on ProFTPD, including configuration and FAQs, is available at
.BR http://www.proftpd.org/
.PP
For help/support, try the ProFTPD mailing lists, detailed on
.BR http://www.proftpd.org/lists.html
.PP
Report bugs at
.BR http://bugs.proftpd.org/


proftpd/NEWS
$Id: NEWS,v 1.594 2004/09/04 18:50:03 castaglia Exp $

1.2.10 - Released 04-Sep-2004
--------------------------------
- Bug 2440 - Unable to use PAM authentication properly.  Use a "*" after
  the module name in an AuthOrder directive to indicate that an auth
  module is authoritative.
- Bug 2441 - AIX5 portability bug with mod_auth_unix, mod_auth_file.
- Bug 2442 - Segfault in FreeBSD PAM library with long login names.
- Bug 2445 - AuthUserFile in <Global> context overrides <VirtualHost> setting.
- Bug 2444 - Use of sendfile() does not interoperate well with RFC2228
  security mechanisms.  Using sendfile(2) to send data bypasses the handling
  of the data by RFC2228 security mechanisms (such as those provided by
  mod_tls).  So if security mechanisms are detected, do not use sendfile().
- Scrub the ScoreboardFile for stale sessions in inetd mode.
- Bug 2427 - proftpd gets a memory fault when run from ssh batch mode.

1.2.10rc3 - Released 13-Jul-2004
--------------------------------
- Fixed typo that prevented 1.2.10rc2 from compiling.

1.2.10rc2 - Released 13-Jul-2004
---------------------------------
- Bug 2396 - NLST command doesn't understand options.  This was caused by
  the solution for Bug 2322.  However, it is not a popular solution, so
  NLST will once again handle options, but only the relevant options.
- Bug 2034 - Add support for a "graceful shutdown" signal. See
  contrib/mod_ctrls_admin.html#shutdown for details.
- Bug 2400 - <Class> search order is wrong.  The documentation correctly
  stated that <Class> sections are matched in order of definition, but the
  code has the match order in the reverse order of definition.
- Bug 2401 - MaxClientsPerClass only checks first directive in config file.
- Bug 2399 - Rename start/stop control actions to up/down.
- Bug 2082 - Add mod_rewrite "replaceall" builtin function.  See the
  RewriteMap documentation for more details.
- Bug 2403 - Sending SIGHUP to proftpd stops it when using Classes.  The fix
  for Bug #2400 could result in an infinite loop during a SIGHUP.
- Bug 2405 - "LIST *" should not list dotfiles.
- Bug 2366 - Add support for -h list option.
- Bug 2332 - SO_OOBINLINE error after upgrading proftpd from 1.2.6 to 1.2.9.
  This is due mostly to a change in the logging; a check for error values
  and logging of them was added.  The setting of this particular socket option
  has been moved earlier in the session, as it was found that short-lived
  TCP connections, as from monitoring systems, would cause this error.
- Bug 2407 - mod_auth_file does not allow for proper cascading of "end" and
  "set" auth requests.
- Bug 2410 - CreateHome always copies skel directory.
- Bug 2336 - Use of /dev/log on Solaris leads to kernel memory leak.
  ProFTPD's use of the /dev/log device on Solaris was tickling a Solaris
  kernel bug that caused the Solaris kernel to leak memory.
- Added a TimeoutLinger directive to complement the --enable-timeout-linger
  configure option.
- Bug 2125 - -vv command line switch should list versions of modules.
- Bug 2420 - Name field is not escaped before querying database.
  The mod_quotatab module was not properly escaping the name string it
  used when looking up records from SQL databases.
- Bug 2424 - SQLDefaultHomedir overrides column value.
- Bug 2411 - Caching effects cause RNTO to fail if AllowOverwrite is off and
  target path does not exist.
- Bug 2422 - %v not working in SQLNamedQuery.
- Bug 2418 - chmod returns 550 with filename containing multiple spaces.
- Bug 2431 - mod_sql does not use UID/GID properly in cache lookups.
- Bug 2303 - Problem evaluating multiple <Class> rules.
- Bug 2419 - Ability to disable TLSRequired on per-user basis (e.g. for
  anonymous logins).
- Bug 2438 - Display variable %z not expanded properly.
- Bug 2439 - <Limit CWD> doesn't work.

1.2.10rc1 - Released 28-Apr-2004
---------------------------------
- Bug 2135 - Add ability to handle passphrase-protected server keys.  mod_tls
  can now properly prompt for passphrases for protected server certificate
  keys when the daemon is starting up.
- Bug 2086 - Add limits for PORT, PASV.  This means that now one can use
  <Limit> to place access controls on the PORT and PASV commands.  This
  applies to the EPRT and EPSV commands as well.
- Bug 2174 - mod_auth_unix should not act authoritatively.  This was causing
  problems when using mod_auth_unix.c and the AuthOrder configuration
  directive.
- Bug 2098 - Added SetEnv and UnsetEnv configuration directives.
- Bug 2271 - Improper autoconf check for getaddrinfo() on Tru64 UNIX 5.1.
  The getaddrinfo symbol is a macro, not a function, on that platform.
- Bug 2255 - RADIUS Service-Type should reflect attribute expectations.
- Added Event API.
- Bug 2272 - Address/port collision check needs to handle DefaultAddress.
- Bug 2072 - Add Controls API.
  This API includes a new program, ftpdctl, that is used to communicate
  directly with the proftpd daemon via a new core module, mod_ctrls.  For
  this new functionality to be used, proftpd must be configured using the
  added --enable-ctrls option.
- Bug 2015 - Add AND, OR keywords to Allow/DenyUser directives.
  The AllowUser, DenyUser, AllowGroup, and DenyGroup directives now take
  an optional keyword that indicates what type of expression they are:
  AND, OR, or regex.  By default, AllowUser and DenyUser are OR expressions,
  and AllowGroup and DenyGroup are AND expressions.  For example:

    AllowUser regex ^ftp
    DenyUser AND dave,bob
    AllowGroup OR web,doc

  These demonstrate that the optional keyword modifier must be the first
  parameter in the configuration directive.
- Bug 2046 - Change RFNR and RNTO logging class to WRITE.  This means that
  ExtendedLogs that use the WRITE logging class will now include the
  RNFR and RNTO commands.
- Mac OS X 10.3 portability fixes.
- Bug 2274 - Default server only binds to one IP address of host if
  the --enable-ipv6 configure option is used.  ProFTPD will now properly

proftpd/utils/ftpshut.c
/*
 * ProFTPD - FTP server daemon
 * Copyright (c) 1997, 1998 Public Flood Software
 * Copyright (c) 2001, 2002, 2003 The ProFTPD Project team
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 *
 * As a special exemption, Public Flood Software/MacGyver aka Habeeb J. Dihu
 * and other respective copyright holders give permission to link this program
 * with OpenSSL, and distribute the resulting executable, without including
 * the source code for OpenSSL in the source distribution.
 */

/* Simple utility to create the proftpd shutdown message file, allowing
 * an admin to configure the shutdown, deny, disconnect times and messages.
 *
 * $Id: ftpshut.c,v 1.7 2003/09/09 14:58:34 castaglia Exp $
 */

#include "conf.h"

static void show_usage(char *progname) {
  printf("usage: %s [ -R ] [ -l min ] [ -d min ] time "
    " [ warning-message ... ]\n", progname);

  exit(1);
}

static int isnumeric(char *str) {
  while (str && isspace((int) *str))
    str++;

  if (!str || !*str)
    return 0;

  for (; str && *str; str++) {
    if (!isdigit((int) *str))
      return 0;
  }

  return 1;
}

int main(int argc, char *argv[]) {
  int deny = 10,disc = 5,c;
  FILE *outf;
  char *shut,*msg,*progname = argv[0];
  time_t now;
  struct tm *tm;
  int mn = 0,hr = 0;

  opterr = 0;

  while ((c = getopt(argc, argv, "Rl:d:")) != -1) {
    switch (c) {
      case 'R':
        if (unlink(SHUTMSG_PATH) < 0) {
          fprintf(stderr, "%s: error removing '" SHUTMSG_PATH "': %s\n",
            progname, strerror(errno));
          exit(1);
        }
        fprintf(stdout, "%s: " SHUTMSG_PATH " removed\n", progname);
        exit(0);

      case 'l':
      case 'd':
        if (!optarg) {
          fprintf(stderr, "%s: -%c requires an argument\n", progname, c);
          show_usage(progname);
        }

        if (!isnumeric(optarg)) {
	  fprintf(stderr, "%s: -%c requires a numeric argument\n", progname, c);
	  show_usage(progname);
        }

        if (c == 'd')
	  disc = atoi(optarg);

        else if (c == 'l')
	  deny = atoi(optarg);

        break;

      case '?':
        fprintf(stderr, "%s: unknown option '%c'\n", progname, (char)optopt);

      case 'h':
      default:
        show_usage(progname);
    }
  }

  /* Everything left on the command line is the message */
  if (optind >= argc)
    show_usage(progname);

  shut = argv[optind++];

  if(optind < argc)
    msg = argv[optind];
  else
    msg = "going down at %s";

  time(&now);
  tm = localtime(&now);

  /* shut must be either 'now', '+number' or 'HHMM' */
  if(strcasecmp(shut,"now") != 0) {
    if(*shut == '+') {
      shut++;
      while (shut && *shut && isspace((int) *shut)) shut++;

      if(!isnumeric(shut)) {
	fprintf(stderr, "%s: Invalid time interval specified.\n", progname);
	show_usage(progname);
      }

      now += (60 * atoi(shut));
      tm = localtime(&now);
    } else {
      if((strlen(shut) != 4 && strlen(shut) != 2) || !isnumeric(shut)) {
	fprintf(stderr, "%s: Invalid time interval specified.\n", progname);
	show_usage(progname);
      }

      if(strlen(shut) > 2) {
        mn = atoi((shut + strlen(shut) - 2));

	if(mn > 59) {
	  fprintf(stderr, "%s: Invalid time interval specified.\n",
		  progname);
	  show_usage(progname);
	}

        *(shut + strlen(shut) - 2) = '\0';
      }

      hr = atoi(shut);

      if(hr > 23) {
	fprintf(stderr, "%s: Invalid time interval specified.\n",
		progname);
	show_usage(progname);
      }

      if(hr < tm->tm_hour || (hr == tm->tm_hour && mn <= tm->tm_min)) {
        now += 86400;		/* one day forward */
        tm = localtime(&now);
      }
      tm->tm_hour = hr;
      tm->tm_min = mn;
    }
  }

  umask(022);
  if ((outf = fopen(SHUTMSG_PATH, "w")) == NULL) {
    fprintf(stderr,"%s: error opening '" SHUTMSG_PATH "': %s\n", progname,
      strerror(errno));
    exit(1);
  }

  fprintf(outf,"%d %d %d %d %d %d",
          tm->tm_year+1900, tm->tm_mon + 1, tm->tm_mday, tm->tm_hour,
          tm->tm_min, tm->tm_sec);
  fprintf(outf," %02d%02d %02d%02d\n",
          (deny / 60),(deny % 60),
          (disc / 60),(disc % 60));
  fprintf(outf,"%s\n",msg);
  fclose(outf);
  return 0;
}
